common/StringVector.hpp

```diff
@@ -111,64 +128,84 @@ class StringVector
     {
         if (size == 0 || data == nullptr || *data == '\0')
             return StringVector();
 
         std::vector<StringToken> tokens;
-        tokenize(data, size, delimiter, tokens);
+        tokenize(data, size, delimiter, tokens, ignoreQuotes);
         return StringVector(std::string(data, size), std::move(tokens));
     }
 
     /// Tokenize single-char delimited values until we hit new-line or the end.
-    static StringVector tokenize(const std::string& s, const char delimiter = ' ')
+    static StringVector tokenize(const std::string& s, const char delimiter = ' ', bool ignoreQuotes = true)
     {
         if (s.empty())
             return StringVector();
 
         std::vector<StringToken> tokens;
-        tokenize(s.data(), s.size(), delimiter, tokens);
+        tokenize(s.data(), s.size(), delimiter, tokens, ignoreQuotes);
         return StringVector(s, std::move(tokens));
     }
 
     /// Tokenize by the delimiter string.
-    static StringVector tokenize(const std::string& s, const char* delimiter, int len = -1)
+    /// Ignoring delimiters that are within quotation marks, unless ignoreQuotes is set to true
+    static StringVector tokenize(const std::string& s, const char* delimiter, int len = -1, bool ignoreQuotes = true)
     {
         if (s.empty() || len == 0 || delimiter == nullptr || *delimiter == '\0')
             return StringVector();
 
         if (len < 0)
             len = std::strlen(delimiter);
 
-        std::size_t start = 0;
-        std::size_t end = s.find(delimiter, start);
+        std::string delimiterString(delimiter);
+        if(delimiterString.find('\'') != std::string::npos || delimiterString.find('"') != std::string::npos)
+            ignoreQuotes = true;
+
+        // need to track two start indexes:
+        // - one for where we have looked for delimiters(in and out of brackets)
+        // - and another for where to slice the string into tokens
+        std::size_t searchStart = 0;
+        std::size_t tokenStart = 0;
+        std::size_t end = 0;
+
+        int singleQuotesBefore = 0;
+        int doubleQuotesBefore = 0;
 
         std::vector<StringToken> tokens;
         tokens.reserve(16);
 
-        tokens.emplace_back(start, end - start);
-        start = end + len;
-
         while (end != std::string::npos)
         {
-            end = s.find(delimiter, start);
-            tokens.emplace_back(start, end - start);
-            start = end + len;
+            end = s.find(delimiter, searchStart);
+
+            if (end != std::string::npos) {
+                // s.begin() + std::string::npos causes issues
+                singleQuotesBefore = std::count(s.begin(), s.begin() + end, '\'');
+                doubleQuotesBefore = std::count(s.begin(), s.begin() + end, '"');
+            }
+            // if no. quotes before the delimiter is odd then it is within quotes
+            if (ignoreQuotes || (singleQuotesBefore % 2 == 0 && doubleQuotesBefore % 2 == 0) ||
```

Comment:
A better algorithm is to find either the delimiter or a quote. If the quote is before the delimiter, then search for the matching quote before searching for the delimiter again, or the end of string. There shouldn't be a need to count quotes (note that counting is a side-effect of the unnecessary quadratic algorithm).